### ISSUE TITLE
experiment: test activation-fitted INT4 scales against absmax and weight-LS at matched bytes

### ISSUE THREAD

**Issue by rmems** [rmems/grok-ozempic#129]
## Planning attribution and dependencies

Planned by **ChatGPT (OpenAI), 2026-09-08**, at the owner's request. This is a **testable hypothesis**, not an implemented quantizer, measured improvement, or claim of novelty.

**Third experimental milestone.** Depends on #125's scale/schedule diagnosis and #127's ordered, disjoint calibration/held-out input contract. Keep #85 closed and its canonical three-arm evidence unchanged. Do not start a broad AWQ/GPTQ integration or group-size sweep in this issue.

## Why test a different objective?

The existing `int4_ls_channel_alpha_scale` minimizes weight reconstruction error for fixed INT4 codes, using `alpha = sum(W * Q) / sum(Q * Q)` per output channel \[S1\]. Nevertheless, the completed #85 all-block alpha arm has worse b3 top-1 agreement and chain-exit drift than absmax INT4 \[S2\]. That does not prove an implementation bug: minimizing weight-space error does not mathematically guarantee minimizing activation/output error or preserving later routing.

AWQ provides a primary-source motivation for using activation distributions rather than weight magnitude alone \[S3\]. GPTQ is another relevant low-bit baseline family using approximate second-order information \[S4\]. **The small fixed-code experiment below is neither AWQ nor GPTQ**, and those papers do not establish that it will work on this Grok-1 chain. It can overfit calibration inputs; held-out evaluation is required.

## One bounded hypothesis

> Holding INT4 codes, per-output-channel scale count/dtype, and expert precision schedule fixed, can scales fitted to calibration projection outputs improve held-out block fidelity and downstream routing over both absmax and weight-only LS scales?

For one expert projection, write its row-major forward operation as `Y = X @ W`, with fixed integer codes `Q` and `Z = X @ Q`. The unconstrained least-squares scalar for output channel j is:

```text
alpha[j] = sum_t(Z[t,j] * Y[t,j]) / sum_t(Z[t,j]^2)
```

This equation is a proposed local-output fitting objective, not an end-to-end routing guarantee. Use the actual tensor orientation and the actual input to each gate/up/down projection; the down projection's input is not the pre-GeGLU hidden state. Stream sufficient statistics in float64 over bounded token/weight chunks rather than storing a giant activation dataset or full Hessian.

Predeclare handling of zero/ill-conditioned denominators and non-finite/non-positive proposed scales. A conservative first implementation retains the original absmax scale in those cases and records why. Do not invent fitted scales for an expert that received no calibration tokens. Report routed token coverage per expert and projection, fallback counts, scale distributions, and local fit error.

## Stage 1 — isolate the fitting objective

Using the same reference calibration trajectory and code payloads, build exactly three scale modes:

1. absmax;
2. existing weight-LS alpha;
3. proposed activation-output LS alpha.

Begin with **INT4 experts on block 0 and FP16 experts on blocks 1–3** for all three arms. This is a sensitivity/isolation experiment, not a candidate with demonstrated whole-model compression value. All nonexpert precision remains fixed and high precision.

Evaluate paired 8192-token held-out windows from #127, seed `20260806`, blocks 0–3, `top_k=2`, with the same high-precision reference and mandatory FP16 control. Freeze the experiment protocol before inspecting candidate held-out metrics. If prior development already used those windows to tune this method, designate them validation data and reserve new disjoint confirmation windows; do not relabel a repeatedly tuned set as untouched test data.

Do not fit on held-out activations, select fallback policy from held-out wins, or silently introduce learned biases, low-rank corrections, changed codes, extra HP blocks, groupwise scales or a new routing policy.

## Stage 2 — conditional, not an automatic larger run

Only after Stage 1 shows a prespecified useful held-out effect without failing controls, repeat the three-way comparison with INT4 experts on **all four blocks**, fitting each block from the calibration-only reference trajectory. Reuse the same method; keep self-propagated residuals in every evaluation arm.

Before any real-weight run, commit the minimum-effect/non-regression gates and a finite run budget. The final report must show b3 and per-block top-1/top-2, router-margin-stratified flips, MoE/projection/output cosine and relative error, chain-exit drift, per-window variability, and FP16 control results. Calibration MSE improvement alone is not a pass. If the method fails to improve held-out fidelity, publish that negative result and stop this issue rather than silently expanding to another algorithm. A separately justified groupwise INT4 or GPTQ-style experiment can be considered afterward.

## Matched cost and provenance

* Keep code arrays identical across the three scale modes and prove that with content hashes. Keep one scale of the same dtype per existing output channel; record actual byte totals, not only the word INT4.
* Bind the new side-table generation to reference weights, q contents, scale bytes, algorithm/version, **calibration corpus/tokenizer/token-order hashes**, and fallback/coverage policy. Weight-only cache identity is insufficient for activation-fitted scales.
* Store the new method in a separate namespace; it must never be accepted as an old absmax or weight-LS cache. Preserve rmems/grok-ozempic#89's content-binding, locks, durability, bounded approximately 64 MiB weight construction and crash/restart semantics.
* Report actual research-cache/scale/FP16 bytes, theoretical nibble-packed bytes separately, fitting cost, wall time, and process-tree memory. No GOZ1 INT4 packaging or inference-speedup claim.

## Acceptance

- [ ] Small fixtures validate the analytical scalar against direct least-squares/output-MSE calculations, including orientation and GeGLU/down-input correctness.
- [ ] Tests cover zero-energy channels, non-finite or invalid fits, unvisited experts, fallback provenance, and calibration leakage rejection.
- [ ] Cache tests reject wrong calibration/tokenizer identity, wrong code/scale content, mixed method namespaces, and incomplete publication; weight-only old methods retain existing behavior.
- [ ] Same-code and same-scale-byte assertions hold across all three modes; expected arm schedules are validated rather than conflated.
- [ ] Stage 1 completes with held-out measurements and controls, or records an explicit inconclusive result with failure artifacts. Stage 2 runs only when its frozen gate is met.
- [ ] Publish one bounded conclusion: helpful at matched cost, local-only gain without downstream benefit, no meaningful improvement/regression, or inconclusive. Do not require a positive result to close the scientific question.

## Local execution constraints / non-goals

Inherit #125's **≥24 GiB MemAvailable, ≥2 GiB free swap, ≥40 GiB free disk plus calculated extra output capacity**, no competing heavyweight job, one child at a time, and no 2048 fallback. Preserve failed runs. No full-model training, distillation/export under #84, SNN/FPGA work, attention/router/norm quantization, 64-block generation, generic quantizer framework, CUDA kernels, or model-family expansion.

## Sources

* **\[S1\]** [Existing weight-space channel-alpha implementation](<https://github.com/rmems/grok-ozempic/blob/a9e9860cb0b2fbec113cf051bf226e06dbc64237/scripts/grok1_multiblock_lib.py>), function `int4_ls_channel_alpha_scale`.
* **\[S2\]** [Canonical #85 measurements and protocol](<https://github.com/rmems/grok-ozempic/blob/972145ee2aabfa88cbaf3a000316b8111a62fc89/reports/grok-1-expert-precision-remedy-v4/results.md>).
* **\[S3\]** Ji Lin et al., [AWQ: Activation-aware Weight Quantization for LLM Compression and Acceleration](<https://arxiv.org/abs/2306.00978>), MLSys 2024; arXiv version 6 available at review. Motivation only; this issue does not implement AWQ's equivalent channel transformation.
* **\[S4\]** Elias Frantar et al., [GPTQ: Accurate Post-Training Quantization for Generative Pre-trained Transformers](<https://arxiv.org/abs/2210.17323>), ICLR 2023. Reference baseline family, not evidence about this custom Grok experiment.
* **\[S5\]** [#125: scale/schedule attribution](<https://github.com/rmems/grok-ozempic/issues/125>), [#127: input and held-out validation contract](<https://github.com/rmems/grok-ozempic/issues/127>), and [PR #89: bounded/content-bound side tables](<https://github.com/rmems/grok-ozempic/pull/89>).

GitHub is canonical; avoid a manually duplicated Linear task.

## Research audit reference

Owner-approved prioritization — 2026-09-21, Codex (OpenAI). Source: [Grok-1 weights and artifacts audit — 8 September 2026](<https://linear.app/rpd-34/document/grok-1-weights-and-artifacts-audit-8-september-2026-1820c6357b9a>). Research sequence: #125 matched scale/schedule controls; #127 ordered held-out text; #129 activation-fitted scales. Existing acceptance criteria and native dependencies remain authoritative. A harness-only PR is not completed experimental evidence.
